libethereum/State.cpp

```diff
@@ -288,126 +297,116 @@ void State::setRoot(h256 const& _r)
 
 bool State::addressInUse(Address const& _id) const
 {
-	ensureCached(_id, false, false);
-	auto it = m_cache.find(_id);
-	if (it == m_cache.end())
-		return false;
-	return true;
+	return !!account(_id);
 }
 
 bool State::addressHasCode(Address const& _id) const
 {
-	ensureCached(_id, false, false);
-	auto it = m_cache.find(_id);
-	if (it == m_cache.end())
+	if (auto a = account(_id))
+		return a->isFreshCode() || a->codeHash() != EmptySHA3;
+	else
 		return false;
-	return it->second.isFreshCode() || it->second.codeHash() != EmptySHA3;
 }
 
 u256 State::balance(Address const& _id) const
 {
-	ensureCached(_id, false, false);
-	auto it = m_cache.find(_id);
-	if (it == m_cache.end())
+	if (auto a = account(_id))
+		return a->balance();
+	else
 		return 0;
-	return it->second.balance();
 }
 
 void State::noteSending(Address const& _id)
 {
-	ensureCached(_id, false, false);
-	auto it = m_cache.find(_id);
-	if (asserts(it != m_cache.end()))
+	Account* a = account(_id);
+	if (asserts(a != nullptr))
 	{
 		cwarn << "Sending from non-existant account. How did it pay!?!";
 		// this is impossible. but we'll continue regardless...
 		m_cache[_id] = Account(requireAccountStartNonce() + 1, 0);
 	}
 	else
-		it->second.incNonce();
+		a->incNonce();
 }
 
 void State::addBalance(Address const& _id, u256 const& _amount)
 {
-	ensureCached(_id, false, false);
-	auto it = m_cache.find(_id);
-	if (it == m_cache.end())
-		m_cache[_id] = Account(requireAccountStartNonce(), _amount, Account::NormalCreation);
+	if (Account* a = account(_id))
+		a->addBalance(_amount);
 	else
-		it->second.addBalance(_amount);
+		m_cache[_id] = Account(requireAccountStartNonce(), _amount, Account::NormalCreation);
 }
 
 void State::subBalance(Address const& _id, bigint const& _amount)
 {
-	ensureCached(_id, false, false);
-	auto it = m_cache.find(_id);
-	if (it == m_cache.end() || (bigint)it->second.balance() < _amount)
+	Account* a = account(_id);
+	if (!a || a->balance() < _amount)
 		BOOST_THROW_EXCEPTION(NotEnoughCash());
 	else
-		it->second.addBalance(-_amount);
+		a->addBalance(-_amount);
 }
 
 void State::createContract(Address const& _address)
 {
 	m_cache[_address] = Account(requireAccountStartNonce(), balance(_address), Account::ContractConception);
 }
 
-void State::kill(Address _a)
+void State::ensureAccountExists(const Address& _address)
+{
+	if (!addressInUse(_address))
```

Comment:
It is confusing for me that we have: "exists", "is empty", and "in use" queries. My guess is "in use" should go away :).